library/src/blas2/rocblas_tbsv.hpp

```diff
@@ -0,0 +1,318 @@
+            if(tx + i < n)
+                xshared[tx] = x[(tx + i) * incx];
+
+            __syncthreads();
+
+            // iterate through the current block and solve elements
+            for(rocblas_int j = 0; j < BLK_SIZE; j++)
+            {
+                // If the current block covers more than what's left,
+                // we break early.
+                if(j + i >= n)
+                    break;
+
+                // solve element that can be solved
+                if(tx == j && !diag)
+                {
+                    rocblas_int colA   = j + i;
+                    rocblas_int rowA   = j + i;
+                    rocblas_int indexA = banded_matrix_index(trans, trans, n, lda, k, rowA, colA);
+                    xshared[tx]        = xshared[tx] / (CONJ ? conj(A[indexA]) : A[indexA]);
+                }
+
+                __syncthreads();
+
+                // for rest of block, subtract previous solved part
+                if(tx > j)
+                {
+                    rocblas_int colA   = j + i;
+                    rocblas_int rowA   = tx + i;
+                    rocblas_int indexA = banded_matrix_index(trans, trans, n, lda, k, rowA, colA);
+
+                    // Ensure row is in range, and subtract
+                    if(rowA < n && colA >= rowA - k)
+                        xshared[tx] -= (CONJ ? conj(A[indexA]) : A[indexA]) * xshared[j];
+                }
+            }
+
+            __syncthreads();
+
+            // apply solved diagonal block to the rest of the array
+            // 1. Iterate down rows
+            for(rocblas_int j = BLK_SIZE + i; j < n; j += BLK_SIZE)
+            {
+                if(tx + j >= n)
+                    break;
+
+                // 2. Sum result (across columns) to be subtracted from original value
+                T val = 0;
+                for(rocblas_int p = 0; p < BLK_SIZE; p++)
+                {
+                    rocblas_int colA   = i + p;
+                    rocblas_int rowA   = tx + j;
+                    rocblas_int indexA = banded_matrix_index(trans, trans, n, lda, k, rowA, colA);
+
+                    if(diag && colA == rowA)
+                        val += xshared[p];
+                    else if(colA < n && colA >= rowA - k)
+                        val += (CONJ ? conj(A[indexA]) : A[indexA]) * xshared[p];
+                }
+
+                x[(tx + j) * incx] -= val;
+            }
+
+            // store solved part back to global memory
+            if(tx + i < n)
+                x[(tx + i) * incx] = xshared[tx];
+
+            __syncthreads();
+        }
+    }
+
+    // Uses backward substitution to solve Ax = b. Used for a non-transposed upper-triangular matrix
+    // or a transposed lower-triangular matrix.
+    template <bool CONJ, rocblas_int BLK_SIZE, typename T>
+    __device__ void tbsv_backward_substitution_calc(
+        bool diag, bool trans, int n, int k, const T* A, rocblas_int lda, T* x, rocblas_int incx)
```

Comment:
same as earlier function trans as a template parameter may help with loop unrolling by reducing the amount of code